src/hotspot/cpu/x86/x86_64.ad

```diff
@@ -5333,9 +5333,9 @@ instruct loadNKlassCompactHeaders(rRegN dst, indOffset8 mem, rFlagsReg cr)
   ins_cost(125); // XXX
   format %{ "movl    $dst, $mem\t# compressed klass ptr" %}
   ins_encode %{
-    assert($mem$$disp == TypeOopPtr::klass_offset_in_bytes(), "expect correct offset 4, but got: %d", $mem$$disp);
-    assert($mem$$index == 4, "expect no index register: %d", $mem$$index);
-    __ load_nklass_compact_c2($dst$$Register, $mem$$base$$Register);
+    Register index = $mem$$index != 4 ? $mem$$index$$Register : noreg;
```

Comment:
I don't understand what`$mem$$index != 4` is supposed to check. Can't we just pass the whole register to `load_nklass_compact_c2`? Is it already `noreg` when index is not set?

Reply:
Or is unset index `-1`, not `4`?

Reply:
All the operand definitions use 0x4 for 'no index' e.g.:
operand stackSlotI(sRegI reg)
%{
  constraint(ALLOC_IN_RC(stack_slots));
  // No match rule because this operand is only generated in matching

  format %{ "[$reg]" %}
  interface(MEMORY_INTER) %{
    base(0x4);   // RSP
    index(0x4);  // No Index
    scale(0x0);  // No Scale
    disp($reg);  // Stack Offset
  %}
%}

There are many more cases like this in x86_64

Reply:
On the other hand, this suggests that we can pass `$mem$$index$Register` right away?

https://github.com/openjdk/lilliput/pull/111/files#diff-018aa61d1a7aafcf70a535fcd40a318a4bd6511fd40ac39ce4be90cc52216749L7350

Reply:
In fact, I don't quite see why `x86_64.ad` cannot do what `aarch64.ad` does:

```
__ load_nklass_compact($dst$$Register, $mem$$base$$Register, $mem$$index$$Register, $mem$$scale, $mem$$disp);
```

Reply:
Problem is the index=4 does not decode to noreg (it decodes to rsp) and scale=0 does not decode to no_scale (it's invalid, I think). Hence the extra hoops there.

Reply:
Ooof. Okay then.